### PR TITLE
Jetty 12.0.x 12191 debug listener

### LINKED ISSUE #12191: DebugListener module in `core` refers to non-existent `org.eclipse.jetty.server.DebugListener`
**Jetty version(s)**
jetty-12.0.x

**Jetty Environment**
core

The `debug.mod` `core` module refers to `jetty-debug.xml`, which refers to `org.eclipse.jetty.server.DebugListener` that does not exist, although `ee` specific `DebugListener` classes do:

```
find . -name DebugListener.java

./jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/DebugListener.java
./jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/DebugListener.java
./jetty-ee8/jetty-ee8-nested/target/generated-sources/ee8/org/eclipse/jetty/ee8/nested/DebugListener.java


find . -name "*.xml" | xargs egrep DebugListener

./jetty-core/jetty-server/src/main/config/etc/jetty-debug.xml:<!-- The DebugListener                                               -->
./jetty-core/jetty-server/src/main/config/etc/jetty-debug.xml:  <New id="DebugListener" class="org.eclipse.jetty.server.DebugListener">
./jetty-core/jetty-server/src/main/config/etc/jetty-debug.xml:  <Call name="addBean"><Arg><Ref refid="DebugListener"/></Arg></Call>
./jetty-core/jetty-server/src/main/config/etc/jetty-debug.xml:        <New class="org.eclipse.jetty.deploy.bindings.DebugListenerBinding">
./jetty-core/jetty-server/src/main/config/etc/jetty-debug.xml:          <Arg><Ref refid="DebugListener"/></Arg>


find . -name debug.mod

./jetty-core/jetty-server/src/main/config/modules/debug.mod

```

### PR DESCRIPTION
**Pull request by janbartel**
Closes #12191

This PR is to re-enable the Debug logging feature. 

In order to do it, I had to extend the concept of the environment-specific context xml files such that we could define any number of `eeX-` properties files, naming xml files to be applied to the webapp being deployed.

Prior to this change, you were only permitted to define a single `eeX.properties` file, containing a single `org.eclipse.jetty.deploy.environmentXml` property pointing to either an absolute filename, or a filename relative to `$jetty.base`.

After this change, you may have any number of properties files prefixed by `eeX-`.  Each one contains a property prefixed with `org.eclipse.jetty.deploy.environmentXml` that names a location of an absolute or `$jetty.base` relative file.

When there are multiple environment-specific xml files they are applied to the webapp being deployed _in the ordering produced by the property names_.  So properties `org.eclipse.jetty.deploy.environmentXml.foo` and `org.eclipse.jetty.deploy.environmentXml.bar` properties will result in their context xml files being applied in the ordering:

1. `org.eclipse.jetty.deploy.environmentXml.bar` 
2. `org.eclipse.jetty.deploy.environmentXml.foo`

Having done the above, we can now deploy the EE specific Debug logging feature. There are modules for `ee8-debug`, `ee9-debug`, `ee10-debug` (and will be `ee11-debug` in jetty-12.1).

Deployment documentation has also been updated to reflect the above.

### CODE REVIEW

**Comment by joakime** [comment]
@olamy huh? what does this mean from https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-12231/2/pipeline ?

```
[INFO] -------------------< org.eclipse.jetty:jetty-maven >--------------------
[INFO] Building Core :: Maven 12.0.13-SNAPSHOT                         [61/352]
[INFO]   from jetty-core/jetty-maven/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
Creating placeholder flownodes because failed loading originals.
ERROR: Cannot resume build because FlowNode 70 for FlowHead 1 could not be loaded. This is expected to happen when using the PERFORMANCE_OPTIMIZED durability setting and Jenkins is not shut down cleanly. Consider investigating to understand if Jenkins was not shut down cleanly or switching to the MAX_SURVIVABILITY durability setting which should prevent this issue in most cases.
[GitHub Checks] GitHub check (name: Jenkins, status: completed) has been published.
```

File: documentation/jetty/modules/operations-guide/pages/deploy/index.adoc [hunk 1]
```diff
@@ -202,40 +202,92 @@ $ java -jar $JETTY_HOME/start.jar -Dmyapps.dir=/opt/myapps
...
 == Deploying Environment Specific Context XML Files
 
-A xref:xml/index.adoc[Jetty context XML file] file can be applied to _all_ webapps deployed to a particular `environment` by adding a file named `{ee-all}.properties` containing the location of the file.
+xref:xml/index.adoc[Jetty context XML file] files can be applied to _all_ webapps deployed to a particular `environment`.
 
-This environment-specific Jetty context XML file will be applied to the webapps _before_ any context XML file associated with the webapp.
-Thus, it can be used to configure general defaults to all webapps for a given `environment`, allowing a webapp-specific context XML file to further refine or override that configuration.
+Add a properties file to the deployment directory that contains a property naming the location of the xml file to apply.
+The properties file name must be prefixed by the name of the `environment`.
+For example `ee8.properties, ee8-more.properties ee8-other.properties` would all apply to the `ee8` environment.
```
Comment (gregw) [inline documentation/jetty/modules/operations-guide/pages/deploy/index.adoc]:
```suggestion
For example `ee8.properties, ee8-more.properties ee8-other.properties` would all apply to all contexts deployed in the `ee8` environment.
```

File: documentation/jetty/modules/operations-guide/pages/deploy/index.adoc [hunk 2]
```diff
@@ -202,40 +202,92 @@ $ java -jar $JETTY_HOME/start.jar -Dmyapps.dir=/opt/myapps
...
+
+The properties files contain at minimum one property whose name is prefixed with `org.eclipse.jetty.deploy.environmentXml`.
```
Comment (gregw) [inline documentation/jetty/modules/operations-guide/pages/deploy/index.adoc]:
```suggestion
If the property file(s) contain one or more properties whose names are prefixed with `org.eclipse.jetty.deploy.environmentXml`, then their values are used as additional context XML files to apply to all contexts deployed in the corresponding environment.
```

File: documentation/jetty/modules/operations-guide/pages/deploy/index.adoc [hunk 3]
```diff
@@ -202,40 +202,92 @@ $ java -jar $JETTY_HOME/start.jar -Dmyapps.dir=/opt/myapps
...
+So for example `org.eclipse.jetty.deploy.environmentXml, org.eclipse.jetty.deploy.environmentXml.more, org.eclipse.jetty.deploy.environmentXml.other` are all acceptable as names.
+Each property configures the location of a context xml file to apply to a context when it is being created and deployed.
```
Comment (gregw) [inline documentation/jetty/modules/operations-guide/pages/deploy/index.adoc]:
```suggestion
Each property configures the location of a context XML file to apply to a context when it is being created and deployed.
```

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 4]
```diff
@@ -356,28 +362,77 @@ public ContextHandler createContextHandler(final App app) throws Exception
...
-                    p.stringPropertyNames().forEach(k -> properties.put(k, p.getProperty(k)));
-                }
+                envPropertyFiles = paths.filter(Files::isRegularFile)
+                    .map(p -> parent.relativize(p))
+                    .filter(p ->
+                    {
+                        String name = p.getName(0).toString();
+                        if (!name.endsWith(".properties"))
+                            return false;
+                        if (!name.startsWith(env))
```
Comment (gregw) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: env can be empty string???

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 5]
```diff
@@ -356,28 +362,77 @@ public ContextHandler createContextHandler(final App app) throws Exception
...
 
-                    context = applyXml(context, envXmlPath, env, properties);
+            Map<String, Path> envXmlFilenameMap = new HashMap<>();
+            for (Path file : envPropertyFiles)
+            {
+                Path resolvedFile = parent.resolve(file);
+                if (Files.exists(resolvedFile))
+                {
+                    Properties tmp = new Properties();
+                    try (InputStream stream = Files.newInputStream(resolvedFile))
```
Comment (gregw) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: Aren't these meant to be merged into the single properties file and then that is checed properties with the prefix?

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java (hunk 5 above)
Comment (janbartel) [reply jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: I don't understand your question, if you look at lines 406 and 407 the properties are all added into the general property substitution pool:
 
``` java
                    try (InputStream stream = Files.newInputStream(resolvedFile))
                    {
                        tmp.load(stream);
                        //put each property into our substitution pool
                        tmp.stringPropertyNames().forEach(k -> properties.put(k, tmp.getProperty(k)));
```

**Comment by janbartel** [comment]
@gregw can you also comment on how the `DebugHandler` is used? There's one in `core` and one in each of `ee8` and `ee9`. The one in core is referred to by the `etc/jetty-debuglog.xml` but that itself is referred to by `debuglog.mod` which says its a deprecated module?

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 6]
```diff
@@ -356,28 +362,77 @@ public ContextHandler createContextHandler(final App app) throws Exception
...
+                            if (name.startsWith(Deployable.ENVIRONMENT_XML))
+                            {
+                                //ensure all environment context xml files are absolute paths
+                                Path envXmlPath = Paths.get(entry.getValue().toString());
+                                if (!envXmlPath.isAbsolute())
+                                    envXmlPath = getMonitoredDirResource().getPath().getParent().resolve(envXmlPath);
+                                //accumulate all properties that name environment xml files so they can be ordered
+                                envXmlFilenameMap.put(name, envXmlPath);
+                            }
+                        }
```
Comment (gregw) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: This loop should be done on properties AFTER it has been merged into by all the individual files. You are doing this on each iteration on the tmp file, so you might actually act on a property that is subsequently replaced in a later property file.

**Review by gregw — CHANGES_REQUESTED** [review]

**Comment by gregw** [comment]
> @gregw can you also comment on how the `DebugHandler` is used? There's one in `core` and one in each of `ee8` and `ee9`. The one in core is referred to by the `etc/jetty-debuglog.xml` but that itself is referred to by `debuglog.mod` which says its a deprecated module?

The DebugHandler exists in ee8/ee9 for backwards compatibility.  All we need going forward is the Core DebugHandler.    I'm not sure why the debuglog module was deprecated, but I assume it was because it does the same thing as the DebugListener.   But DebugListener can only be used in a servlet context, so we would still need DebugHandler for core applications or to have a single debug log over multiple contexts.      So perhaps remove the deprecation and explain a bit more about how they relate to each other.

**Comment by janbartel** [comment]
I've created PR #12254 as an alternative implementation which uses the `DebugHandler`  class instead of ee-specific  `DebugListeners.`

As it stands with this PR, all debug logs from all contexts append to the same log file, however there is no over-arching locking, which will lead to problems.

Ideally, each context would have its own debug log, but at the time at which debug is configured, nothing identifying about the context is known: we don't know the war file, or the context path for example. So to pursue this approach we would have to use eg the hashcode of the `WebAppContext` instance to make a unique filename, which is not really obvious for users.

The alternate implementation in #12254 inserts only a single `DebugHandler` into the handler list, so thus whilst the log will contain debug from all contexts, as there is only a single writer (with appropriate locking) there are no synchronisation issues. However, we can't log information about the context to which the inbound request will be routed, because the point at which it is intercepted, the decision on routing has not yet been made.

Yet another alternative solution that would allow us to have a debug log per context might be to configure the `DebugListeners` in a `Configuration` instead. This would require a user to configure the name of the debug log file, and any other debug log characteristics as eg context attributes. It would also mean that we would need to move the `DebugListener` classes out of `ee10-webapp`  `ee8/9-nested` and into it's own jar so that it can optionally be included on the classpath and therefore optionally enable debug.

@gregw your thoughts?

**Comment by joakime** [comment]
> Ideally, each context would have its own debug log, but at the time at which debug is configured, nothing identifying about the context is known: we don't know the war file, or the context path for example. So to pursue this approach we would have to use eg the hashcode of the `WebAppContext` instance to make a unique filename, which is not really obvious for users.

We are using slf4j, there's nothing preventing us from using a MDC or NDC to identify the context.
From there, a sifting appender from any slf4j implementation (like logback) could split up the logs into separate log files.

**Comment by sbordet** [comment]
> We are using slf4j, there's nothing preventing us from using a MDC or NDC to identify the context.

MDC is thread local, so it won't work in general.

**Comment by janbartel** [comment]
I'm closing this PR in favour of PR #12254. The problem with this PR is that all contexts will share the same log file and it is difficult to see how to make them use a separate logfile in a way that would be meaningful to users.

**Pull request closed without merging**